crypto/pkcs7/pk7_doit.c

```diff
@@ -1117,6 +1119,10 @@ int PKCS7_signatureVerify(BIO *bio, PKCS7 *p7, PKCS7_SIGNER_INFO *si,
         goto err;
     }
 
+    if (OBJ_obj2nid(si->digest_enc_alg->algorithm) == EVP_PKEY_RSA_PSS) {
+        pkey->type = EVP_PKEY_RSA_PSS;
+    }
+
```

Comment:
This is not the right fix. You should not modify the pkey type. Instead set the padding type for the verify operation. `EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING);` However to be able to do that, EVP_DigestVerify... functions have to be used instead of EVP_Verify.... functions.